### ISSUE TITLE
Singleton or preloaded enums not usable as constant expressions

### ISSUE THREAD

**Issue by menip** [godotengine/godot#14681]
Version: Godot 3, beta 1.

If you have an enum in a script, you are able to use it as an export hint:
`export(MyEnum) var options`

If you have an autoload singleton with an enum, there is an error stating "Expected constant expression":
`export(MySingleton.MyEnum) var options`

Steps to reproduce:
- Create an autoload singleton with an enum
- Try to use that enum as an export hint in another script

Link to minimal example project: [SingletonEnum.zip](https://github.com/godotengine/godot/files/1561380/SingletonEnum.zip)

**Comment by akien-mga** [comment]
CC @bojidar-bg - I'm not sure if singletons can be used at all in constant expression though, they might be registered too late for that.

**Comment by thimenesup** [comment]
It works if you use
`export(preload("res://enum.gd").ENUMERATOR) var enumeration`

**Comment by bojidar-bg** [comment]
It should work as long as `const X = MySingleton.MyEnum` works, since it uses the same way to get the constant value.

**Comment by menip** [comment]
@thimenesup Confirming that this works.

**Comment by menip** [comment]
@bojidar-bg Doesn't work, still requires a preload.

**Comment by menip** [comment]
Am I missing something about how autoload singletons work? I thought that one of their benefits is you could use them in other classes, without having to preload. 
For single values and calling functions, you don't need to preload. Why do you need to do it for these kind of assignments?

**Comment by BeayemX** [comment]
When I'm using `export(preload("res://enum.gd").ENUMERATOR) var enumeration` the dropdown will show up in the inspector but when evaluating the value in the script it is always `null`.

**Comment by rosshadden** [comment]
> Basically the enum only has the correct value as long as the reset-value button is shown.

I was searching issues to see if someone had made an issue for this, and I was otherwise going to.  It's not the main point of this issue so I think it warranted its own issue:  #15585.

**Comment by DevinPentecost** [comment]
Hi,

I'm running into this same issue. I'd like to export an enum defined in a singleton.

I would like to avoid having to preload everything, as that defeats the purpose of singletons.

I can confirm that trying to use
```
const _my_enum = singleton.target_enum
export(_my_enum) var the_enum
```

does not work either, and is still an extra, unnecessary step.

Is this impossible due to the current implementation of singletons?

Thank you,
-Devin

**Comment by vnen** [comment]
I forgot about this. Now that singletons are loaded in-editor this is easy to fix.

**Comment by Jendoliver** [comment]
Happens again on 3.1.1-stable, but it's also possible to work around it with `preload` as @thimenesup suggested. Assigning to a `const` also causes the same error.

![invalid](https://user-images.githubusercontent.com/23485344/65378525-dc06bb80-dcb9-11e9-87df-76b2d0bda64e.gif)

**Comment by thomas-goerlich** [comment]
I just ran into the same problem on the master branch (latest commit 5a23ab61fae79cce576fe7ec60fd928d2bd27fca).

**Comment by williamd1k0** [comment]
> @Jendoliver Happens again on 3.1.1-stable, but it's also possible to work around it with preload as @thimenesup suggested. Assigning to a const also causes the same error.

That's weird, I have similar code and it works fine in 3.1.1-stable (tested in the editor with and without tool mode, and in the exported game).

Minimal example project: [SingletonEnums.zip](https://github.com/godotengine/godot/files/3693997/SingletonEnums.zip)

**Comment by thomas-goerlich** [comment]
> 
> 
> > @Jendoliver Happens again on 3.1.1-stable, but it's also possible to work around it with preload as @thimenesup suggested. Assigning to a const also causes the same error.
> 
> That's weird, I have similar code and it works fine in 3.1.1-stable (tested in the editor with and without tool mode, and in the exported game).
> 
> Minimal example project: [SingletonEnums.zip](https://github.com/godotengine/godot/files/3693997/SingletonEnums.zip)

This code is working for me in 3.1.1 stable as well as on my master branch.

I think i found the problem at least in my case. In my globals file i define an enum and i preload another scene which is using the enum. That does not seem to work. If i do not preload the scene in the globals everything is fine.
My Code:
[GameOfLife.zip](https://github.com/godotengine/godot/files/3694297/GameOfLife.zip)
Same enum lines in game.gd/tile.gd. Working fine in game.gd, error in tile.gd. If preload line is moved from globals.gd to game.gd everything is fine.
From this error i would assume that i can only access the enums ones the whole file is parsed?

**Comment by MrBlackswordsman** [comment]
@thomas-goerlich Can confirm removing the preloaded scenes from the autoloaded script fixed the issue for me in 3.2 alpha 1.

**Comment by rasikrodri** [comment]
This is still happening in version 3.2.2.
Please check again!

**Comment by rasikrodri** [comment]
This is still happening in version 3.2.2.
Please check again!

**Comment by Calinou** [comment]
@rasikrodri Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

**Comment by menip** [comment]
@rasikrodri I just ran the minimal reproduction project linked in original post via 3.2.2, and everything seems fine.

**Comment by artism90** [comment]
`preload()` calls break enums within the same Singleton class on each project reopening. This is still happening in 3.2.3.

**Temporary workaround (also shown in the gif):**
1. Edit the Singleton class (no saving required) and see the constant expression error vanishing in the affected scripts.
1. Edit the broken class (no saving required) to rerun the parser and make export variables visible in the inspector since the script is now valid.

Basically the syntax parser has to revisit all broken scripts to validate them.

![SingletonEnumPreload](https://user-images.githubusercontent.com/2534873/93719475-18961d00-fb83-11ea-9f81-e81263be790e.gif)

**Permanent workaround**
Outsource the enums into a distinct Singleton.

Minimal reproduction project:
[SingletonEnum2.zip](https://github.com/godotengine/godot/files/5252388/SingletonEnum2.zip)

**Comment by chucklepie** [comment]
Hello, this problem just goes from bad to worse, with seemingly no fix.

If you have more than one enum and try to use the preload option, e.g.

file1:
```
export(preload("res://singletons/Globals.gd").PICKUP) var pickup_type
```

file2:
```
export(preload("res://singletons/Globals.gd").WEAPON) var weapon_type
```

On one of the files your game aborts with the error:
E 0:00:00.555   load: Resource: 'res://singletons/Globals.gd' is already being loaded. Cyclic reference?
  <C++ Error>   Condition "!success" is true. Returned: RES()
  <C++ Source>  core/io/resource_loader.cpp:360 @ load()

So you cannot call export(Globals.MYENUM) as you get index errors and you cannot call preload because you get cyclic errors

**Comment by ThaYoost** [comment]
Just confirming this is still an issue in 3.2.3 I've got a enum declared in my singleton (EWeaponTypes) which I'd like to use as an exported dropdown on my Player. 

In Player the following code does not work. 
export(WeaponList.EWeaponTypes) var weapon_type 

Error is "Parser error: invalid index EWeaponTypes in constant expression"

**Comment by chucklepie** [comment]
Well I've scrapped enums in most of my game, going thru become part of
4.0...

If you can't pass them as parameters (casting to int is not the same)
either, there's not much point in them really.

**Comment by hectormartinezdev** [comment]
Confirming that removing preloads from my singleton autoload script fixed it. It was:

```
extends Node

enum INTERACTION_TYPE {TALK = 0, INSPECT, SURPRISE}

var current_scene = null
var TransitionLayer = preload("res://UI/TransitionLayer.tscn")

func _ready() -> void:
	current_scene = get_tree().current_scene
```

And I changed it to:


```
extends Node

enum INTERACTION_TYPE {TALK = 0, INSPECT, SURPRISE}

var current_scene = null
var TransitionLayer = null

func _ready() -> void:
	current_scene = get_tree().current_scene
	TransitionLayer = load("res://UI/TransitionLayer.tscn")
```

And it stopped complaining. For me it is not a big deal doing this change, so I'm going with this.

**Comment by hectorzg** [comment]
I didn't have any preloads on my singleton but what fixed the issue for me was to move my Enums autoload to the very first place in the list.

Project > Project Settings > AutoLoad

Locate your script and move it to the top by clicking the arrows near the end of the row.

**Comment by chucklepie** [comment]
Not much use when you have them across multiple autoload though. Hence the
false 'cyclic' errors you cannot bypass

**Comment by Calinou** [comment]
Can anyone reproduce this on [4.0alpha6](https://godotengine.org/article/dev-snapshot-godot-4-0-alpha-6)?

**Comment by MewPurPur** [comment]
> Can anyone reproduce this on [4.0alpha6](https://godotengine.org/article/dev-snapshot-godot-4-0-alpha-6)?
Does this count?
![image](https://user-images.githubusercontent.com/85438892/163846066-0f495d1b-3156-4c87-afcc-ead3deb51910.png)

**Comment by MewPurPur** [comment]
[Uploading Enum problem.zip…]()

Seems like a half-bug, but anyway here's the minimal project.

**Comment by DaGamingWolf** [comment]
this same error seems to happen when inheriting the enum from a parent class' parent class. if it's in the script itself, or a parent class, then the error doesn't appear. but it it's in a parent's parent, or more, then it does. it seems almost as if the script isn't actually inheriting the enum, yet at the same time can detect the enum in the parent class when writing a version of the enum in the script.

**Comment by tokengamedev** [comment]
I was working on a project and and during the final testing I noticed this error. Backtracking the changes I made, I found out there was a change I made to the globals file which started causing this error.

The reason I had this error because I added preload statements like the one given in the issue.
If you remove the preload statements from the file, then everything is fine.

I also tested with example provided here https://github.com/godotengine/godot/issues/14681#issuecomment-695824393
It worked.

My two cents here:
It may be working as designed.
Because if you you add preload statements to a file, then the file will be loaded later than the static files. which is causing the error to be displayed in the editor, but when the game runs the preload variables are compiled already so no issues faced during game run.
My input will be create different autoload files for various constants and variables as per the project.... my lesson for me will be don't mix preloads with constants

Or I may be wrong here as I am new to the scenario, can somebody in the issue can check out too.

**Comment by dalexeev** [comment]
Fixed in 4.x (probably by #70613 or #69590), now you can:

```gdscript
@export var a: Singleton.MyEnum
@export var b: MyClass.MyEnum
# It doesn't matter if MyClass is a global class (`class_name`)
# or a `const` `preload`ed script.
```

And even:

```gdscript
const SEnum = Singleton.MyEnum
const CEnum = MyClass.MyEnum

@export var c: SEnum
@export var d: CEnum
```

> Does this count?

Was fixed. Note that you may encounter a _similar_ issue if the initializer value is not in the enum, but this is more of an Inspector issue, see https://github.com/godotengine/godot/issues/75555#issuecomment-1492996207. Also this is documented ([one](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_exports.html#exporting-enums), [two](https://docs.godotengine.org/en/stable/classes/class_%40gdscript.html#class-gdscript-annotation-export-enum)).
